README.md
Elasticsearch Indexer
=========
![Elasticsearch Indexer](http://mikael.ninja/github/assets/elasticsearch-indexer/banner-728x236.png)

[![StyleCI](https://styleci.io/repos/34813501/shield?style=flat)](https://styleci.io/repos/34813501)
[![License](https://img.shields.io/packagist/l/wallmanderco/elasticsearch-indexer.svg?style=flat)](https://packagist.org/packages/wallmanderco/elasticsearch-indexer)

A plugin that integrates [WordPress](https://github.com/WordPress/WordPress) with [Elasticsearch](https://www.elastic.co/products/elasticsearch).

Aside from super fast full text search, this plugin will speed up post listings of any post type, including [WooCommerce](http://www.woothemes.com/woocommerce/) products.

The plugin works by indexing all posts and post meta data and redirecting the requests by WordPress from MySQL to Elasticsearch.

The plugin requires that you have Elasticsearch installed on your server.

WordPress Repository: https://wordpress.org/plugins/elasticsearch-indexer/

## Demo
 - [WooCommerce 10 000 products](http://enabled-1.es-demo.wallmanderco.se/)
 - [WooCommerce 10 000 products without elasticsearch](http://disabled-1.es-demo.wallmanderco.se/)
 - [WooCommerce 135 000 products](http://enabled-2.es-demo.wallmanderco.se/)
 - [WooCommerce 135 000 products without elasticsearch](http://disabled-2.es-demo.wallmanderco.se/)

## Installing the plugin from GitHub
 1. Download the [zip](https://github.com/wallmanderco/elasticsearch-indexer/archive/master.zip) file from GitHub.
 1. Unzip and put the new directory inside you plugins directory (`/wp-content/plugins/`) in wordpress.
 1. run `composer install` inside the new directory.
 1. [Install Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/guide/current/_installing_elasticsearch.html) if you haven't already.
 1. Make sure that Elasticsearch is installed and running on your server  
(You can test this by running `curl -XGET -i 'http://127.0.0.1:9200'` in the command line)
 1. Activate the plugin in the wordpress admin panel.
 1. Click on the new menu item and press the button “Reindex all posts”.

## Installing Elasticsearch
Follow the instructions in the [Elasticsearch Docs](https://www.elastic.co/guide/en/elasticsearch/guide/current/_installing_elasticsearch.html).

Make sure that Elasticsearch is installed and running on your server by running `curl -XGET -i 'http://127.0.0.1:9200'` in the command line.

## Usage
The plugin requires no modification to work. Once you have put in your settings and run an initial reindex of the site the plugin will keep elasticsearch in sync with MySQL and redirect queries from MySQL to Elasticsearch. However the plugin comes with a few extra features and filters that you can use.

## Not supported
 - Sticky posts (ignores sticky posts. All wordpress queries will act as if `ignore_sticky_posts` is set to `true`)
 - Password protected posts (fallback to MySQL if used)
 - Random sorting (ignored)

## License
Elasticsearch Indexer is licensed under [General Public License (GPLv2)](LICENSE).


CHANGELOG.md
# CHANGELOG

## 1.6.1 (released 2015-08-11)

- Add compatibility with third party plugins like WPML

## 1.6.0 (released 2015-06-18)

- Add option to only use plugin for searching
- Improved order search

## 1.5.1 (released 2015-06-18)

- Improved order search

## 1.5.0 (released 2015-06-15)

- Improved order search
- Improved product search
- Add versioning on index
- Temporary deactivation of integration when indexing

## 1.4.2 (released 2015-06-11)

- Add support for nested meta_queries

## 1.4.1 (released 2015-06-11)

- Add support for include_children on tax_query
- Bugfix

## 1.4.0 (released 2015-06-09)

- Add support for numeric meta queries
- Bugfix

## 1.3.0 (released 2015-06-09)

- Add indexer status to admin bar
- Disable the integration while indexing
- Add php version check

## 1.2.4 (released 2015-05-31)

- Bug fix

## 1.2.3 (released 2015-05-29)

- Important bug fix
- Progressbar when indexing

## 1.2.2 (released 2015-05-28)

- Fix a problem with indexing very large meta values.
- Add logging system

## 1.2.1 (released 2015-05-23)

- Update search analyzers

## 1.2.0 (released 2015-05-23)

- Add order searching for WooCommerce Admin

## 1.1.3 (released 2015-05-21)

- Add connection information on settings panel

## 1.1.2 (released 2015-05-19)

- Speed optimization
- Add support for https
- Add profiler
- Bug fixes

## 1.1.1 (released 2015-05-18)

- Bug fixes

## 1.1.0 (released 2015-05-16)

- Indexing and speed optimizations
- Bug fixes

## 1.0.0 (released 2015-05-05)

- Add support for taxonomy queries
- Add support for date queries
- Bug fixes

## 0.1.0 (released 2015-04-29)

- Initial beta release


config/mappings.php
<?php

/*
 * This file is part of Elasticsearch Indexer.
 *
 * (c) Wallmander & Co <mikael@wallmanderco.se>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

/**
 * Full Credits to 10up/ElasticPress.
 */
$dateTermFields = [
    //4 digit year (e.g. 2011)
    'year' => [
        'type' => 'integer',
    ],
    //Month number (from 1 to 12) alternate name "monthnum"
    'month' => [
        'type' => 'integer',
    ],
    //YearMonth (For e.g.: 201307)
    'm' => [
        'type' => 'integer',
    ],
    //Week of the year (from 0 to 53) alternate name "w"
    'week' => [
        'type' => 'integer',
    ],
    //Day of the month (from 1 to 31)
    'day' => [
        'type' => 'integer',
    ],
    //Accepts numbers 1-7 (1 is Sunday)
    'dayofweek' => [
        'type' => 'integer',
    ],
    //Accepts numbers 1-7 (1 is Monday)
    'dayofweek_iso' => [
        'type' => 'integer',
    ],
    //Accepts numbers 1-366
    'dayofyear' => [
        'type' => 'integer',
    ],
    //Hour (from 0 to 23)
    'hour' => [
        'type' => 'integer',
    ],
    //Minute (from 0 to 59)
    'minute' => [
        'type' => 'integer',
    ],
    //Second (0 to 59)
    'second' => [
        'type' => 'integer',
    ],
];

return [
    'post' => [
        'date_detection'    => false,
        'dynamic_templates' => [
            [
                'template_meta' => [
                    'path_match' => 'post_meta.*',
                    'mapping'    => [
                        'type'   => 'multi_field',
                        'path'   => 'full',
                        'fields' => [
                            '{name}' => [
                                'type'     => 'string',
                                'index'    => 'analyzed',
                                'analyzer' => 'esi_search_analyzer',
                            ],
                            'raw' => [
                                'type'         => 'string',
                                'index'        => 'not_analyzed',
                                'ignore_above' => 256,
                            ],
                        ],
                    ],
                ],
            ],
            [
                'template_meta_num' => [
                    'path_match' => 'post_meta_num.*',
                    'mapping'    => [
                        'type' => 'long',
                    ],
                ],
            ],
            [
                'template_terms' => [
                    'path_match' => 'terms.*',
                    'mapping'    => [
                        'type'       => 'object',
                        'path'       => 'full',
                        'properties' => [
                            'name' => [
                                'type'     => 'string',
                                'index'    => 'analyzed',
                                'analyzer' => 'esi_search_analyzer',
                            ],
                            'term_id' => [
                                'type' => 'long',
                            ],
                            'parent' => [
                                'type' => 'long',
                            ],
                            'slug' => [
                                'type'  => 'string',
                                'index' => 'not_analyzed',
                            ],
                            'all_slugs' => [
                                'type'  => 'string',
                                'index' => 'not_analyzed',
                            ],
                        ],
                    ],
                ],
            ],
            [
                'term_suggest' => [
                    'path_match' => 'term_suggest_*',
                    'mapping'    => [
                        'type'     => 'completion',
                        'analyzer' => 'esi_search_analyzer',
                    ],
                ],
            ],
        ],
        '_all' => [
            'enabled' => false,
        ],
        'properties' => [
            'post_id' => [
                'type'  => 'long',
                'index' => 'not_analyzed',
            ],
            'post_author' => [
                'type'   => 'object',
                'path'   => 'full',
                'fields' => [
                    'display_name' => [
                        'type'     => 'string',
                        'analyzer' => 'standard',
                    ],
                    'login' => [
                        'type'     => 'string',
                        'analyzer' => 'standard',
                    ],
                    'id' => [
                        'type'  => 'long',
                        'index' => 'not_analyzed',
                    ],
                    'raw' => [
                        'type'  => 'string',
                        'index' => 'not_analyzed',
                    ],
                ],
            ],
            'post_date' => [
                'type'   => 'date',
                'format' => 'YYYY-MM-dd HH:mm:ss',
            ],
            'post_date_gmt' => [
                'type'   => 'date',
                'format' => 'YYYY-MM-dd HH:mm:ss',
            ],
            'post_title' => [
                'type'   => 'multi_field',
                'fields' => [
                    'post_title' => [
                        'type'     => 'string',
                        'analyzer' => 'esi_search_analyzer',
                        'store'    => 'yes',
                    ],
                    'raw' => [
                        'type'  => 'string',
                        'index' => 'not_analyzed',
                    ],
                ],
            ],
            'post_excerpt' => [
                'type' => 'string',
            ],
            'post_content' => [
                'type'     => 'string',
                'analyzer' => 'esi_search_analyzer',
                'index'    => 'analyzed',
            ],
            'post_status' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
            'post_name' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
            'post_modified' => [
                'type'   => 'date',
                'format' => 'YYYY-MM-dd HH:mm:ss',
            ],
            'post_modified_gmt' => [
                'type'   => 'date',
                'format' => 'YYYY-MM-dd HH:mm:ss',
            ],
            'post_parent' => [
                'type'  => 'long',
                'index' => 'not_analyzed',
            ],
            'post_type' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
            'post_mime_type' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
            'permalink' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
            'terms' => [
                'type' => 'object',
            ],
            'post_meta' => [
                'type' => 'object',
            ],
            'post_meta_num' => [
                'type' => 'object',
            ],
            'post_date_object' => [
                'type'   => 'object',
                'path'   => 'full',
                'fields' => $dateTermFields,
            ],
            'post_date_gmt_object' => [
                'type'   => 'object',
                'path'   => 'full',
                'fields' => $dateTermFields,
            ],
            'post_modified_object' => [
                'type'   => 'object',
                'path'   => 'full',
                'fields' => $dateTermFields,
            ],
            'post_modified_gmt_object' => [
                'type'   => 'object',
                'path'   => 'full',
                'fields' => $dateTermFields,
            ],
            'menu_order' => [
                'type'  => 'long',
                'index' => 'not_analyzed',
            ],
            'comment_count' => [
                'type'  => 'long',
                'index' => 'not_analyzed',
            ],
            'guid' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
            'order_item_names' => [
                'type'  => 'string',
                'index' => 'not_analyzed',
            ],
        ],
    ],
];
